src/Rendering/TOC.hs

```diff
@@ -98,54 +100,21 @@ newtype DedupeKey = DedupeKey (T.Text, T.Text) deriving (Eq, Ord)
 -- 2. Two similar entries (defined by a case insensitive comparison of their
 --    identifiers) are in the list.
 --    Action: Combine them into a single Replaced entry.
-dedupe :: [Entry Declaration] -> [Entry Declaration]
-dedupe = let tuples = sortOn fst . Map.elems . snd . foldl' go (0, Map.empty) in (fmap . fmap) snd tuples
-  where
-    go :: (Int, Map.Map DedupeKey (Int, Entry Declaration))
-       -> Entry Declaration
-       -> (Int, Map.Map DedupeKey (Int, Entry Declaration))
-    go (index, m) x | Just (_, similar) <- Map.lookup (dedupeKey x) m
-                    = if exactMatch similar x
-                      then (succ index, m)
-                      else
-                        let replacement = Replaced (entryPayload similar)
-                        in (succ index, Map.insert (dedupeKey replacement) (index, replacement) m)
-                    | otherwise = (succ index, Map.insert (dedupeKey x) (index, x) m)
-
-    dedupeKey entry = DedupeKey (toCategoryName (entryPayload entry), T.toLower (declarationIdentifier (entryPayload entry)))
-    exactMatch = (==) `on` entryPayload
-
--- | Construct a 'TOCSummary' from an 'Entry'.
-entrySummary :: Entry Declaration -> TOCSummary
-entrySummary entry = case entry of
-  Changed  a -> recordSummary "modified" a
-  Deleted  a -> recordSummary "removed" a
-  Inserted a -> recordSummary "added" a
-  Replaced a -> recordSummary "modified" a
+dedupe :: [(Change, Declaration)] -> [(Change, Declaration)]
+dedupe = map ((change :: Dedupe -> Change) &&& decl) . sortOn index . Map.elems . foldl' go Map.empty . zipWith (uncurry . Dedupe) [0..] where
```

Comment:
This line is a little long; I would write it with newlines between invocations of `.`, which gives you some nice space on the page to do FORTH-style comments about what happens at each step of the pipeline.